File: frontend/src/lib/providers/akashml.ts
import type { GenerationProvider, ProviderJobStatus, SubmitInput } from './types';
import { modelRegistry } from '@/lib/models/registry';

// AkashML: managed inference API on Akash Network decentralized GPUs.
// Docs: https://chatapi.akash.network/  (OpenAI-compatible for chat; image/video endpoints TBD)
// TODO: Verify exact endpoint structure for image/video/audio generation once API key is available.

const BASE = process.env.AKASHML_API_URL ?? 'https://api.akash.network/inference/v1';

function headers() {
  const key = process.env.AKASHML_API_KEY;
  if (!key) throw new Error('AKASHML_API_KEY is not set');
  return {
    Authorization: `Bearer ${key}`,
    'Content-Type': 'application/json',
  };
}

type AkashJob = {
  id: string;
  status: 'queued' | 'running' | 'completed' | 'failed';
  outputs?: Array<{ url: string }>;
  error?: string;
  progress?: number;
};

function mapStatus(job: AkashJob, modality: 'image' | 'video' | 'audio'): ProviderJobStatus {
  if (job.status === 'queued' || job.status === 'running') {
    return { status: 'processing', progress: job.progress };
  }
  if (job.status === 'completed') {
    return {
      status: 'completed',
      outputs: (job.outputs ?? []).map((o) => ({ url: o.url, type: modality })),
    };
  }
  return { status: 'failed', error: job.error ?? 'Unknown error' };
}

export class AkashMLProvider implements GenerationProvider {
  readonly name = 'akashml' as const;

  async submit(input: SubmitInput): Promise<{ providerJobId: string }> {
    const model = modelRegistry[input.modelSlug];
    if (!model) throw new Error(`Unknown model: ${input.modelSlug}`);

    const cfg = model.providers.akashml;
    if (!cfg) throw new Error(`Model ${input.modelSlug} has no AkashML config`);

    // TODO: confirm actual submit endpoint + request shape from AkashML docs
    const res = await fetch(`${BASE}/jobs`, {
      method: 'POST',
      headers: headers(),
      body: JSON.stringify({
        model: cfg.modelId,
        inputs: input.params,
        ...(input.webhookUrl ? { webhook_url: input.webhookUrl } : {}),
      }),
    });

    if (!res.ok) {
      const err = await res.text();
      throw new Error(`AkashML submit failed (${res.status}): ${err}`);
    }

    const data = (await res.json()) as AkashJob;
    return { providerJobId: data.id };
  }

  async poll(providerJobId: string): Promise<ProviderJobStatus> {
    // TODO: confirm actual poll endpoint from AkashML docs
    const res = await fetch(`${BASE}/jobs/${providerJobId}`, {
      headers: headers(),
    });

    if (!res.ok) {
      throw new Error(`AkashML poll failed (${res.status})`);
    }

    const data = (await res.json()) as AkashJob;
    return mapStatus(data, 'image'); // caller overrides type from DB modality
  }

  async cancel(providerJobId: string): Promise<void> {
    // TODO: confirm cancel endpoint
    await fetch(`${BASE}/jobs/${providerJobId}/cancel`, {
      method: 'POST',
      headers: headers(),
    });
  }
}

File: backend/providers/akashml.py
import httpx
from django.conf import settings
from .base import BaseProvider, SubmitInput, JobStatus


class AkashMLProvider(BaseProvider):
    name = 'akashml'

    def __init__(self, api_key: str | None = None, api_url: str | None = None):
        self._api_key = api_key or getattr(settings, 'AKASHML_API_KEY', '')
        self._api_url = api_url or getattr(settings, 'AKASHML_API_URL', '')

    def _headers(self) -> dict:
        return {
            'Authorization': f'Bearer {self._api_key}',
            'Content-Type': 'application/json',
        }

    def _base(self) -> str:
        return self._api_url.rstrip('/')

    def submit(self, inp: SubmitInput) -> str:
        from providers.registry import MODEL_REGISTRY
        cfg = MODEL_REGISTRY.get(inp.model_slug, {})
        model_id = cfg.get('providers', {}).get('akashml', {}).get('model_id', inp.model_slug)

        payload = {'model': model_id, 'input': inp.params}
        with httpx.Client(timeout=30) as client:
            r = client.post(f'{self._base()}/predictions', headers=self._headers(), json=payload)
            r.raise_for_status()
            return r.json()['id']

    def poll(self, provider_job_id: str, modality: str = 'image') -> JobStatus:
        with httpx.Client(timeout=15) as client:
            r = client.get(f'{self._base()}/predictions/{provider_job_id}', headers=self._headers())
            r.raise_for_status()
            data = r.json()

        status_map = {'starting': 'processing', 'processing': 'processing',
                      'succeeded': 'completed', 'failed': 'failed'}
        status = status_map.get(data.get('status', ''), 'processing')
        outputs = []
        if status == 'completed' and data.get('output'):
            raw = data['output']
            if isinstance(raw, str):
                raw = [raw]
            outputs = [{'url': u, 'type': modality} for u in raw if isinstance(u, str)]

        return JobStatus(status=status, outputs=outputs, error=data.get('error'))

    def cancel(self, provider_job_id: str) -> None:
        with httpx.Client(timeout=10) as client:
            client.post(f'{self._base()}/predictions/{provider_job_id}/cancel', headers=self._headers())
